In [2]:
pip install psycopg2-binary

Note: you may need to restart the kernel to use updated packages.


In [3]:
import pandas as pd
import datetime
import matplotlib.pyplot as plt
import seaborn as sns
import warnings; warnings.filterwarnings(action='once')

from sqlalchemy import create_engine
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'}, echo=False)

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## **Описание данных**

### Таблица **books**

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

In [4]:
query = '''
Select *
FROM books
;
    '''

pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

**Вывод**: Таблица books обладает 6 столбцами и 1000 строками.

### Таблица **authors**

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

In [5]:
query = '''
Select *
FROM authors
;
    '''

pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

**Вывод**: Таблица authors состоит из  2 столбцов и 636 строк.

### Таблица **publishers**

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

In [6]:
query = '''
Select *
FROM publishers
    '''

pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

**Вывод**: Таблица authors состоит из  2 столбцов и 340 строк.

### Таблица **ratings**

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

In [7]:
query = '''
Select *
FROM ratings
    '''

pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

**Вывод**: Таблица ratings состоит из  4 столбцов и 6456 строк.

### Таблица **reviews**

Содержит данные о пользовательских обзорах на книги:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

In [8]:
query = '''
Select *
FROM reviews
    '''

pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

**Вывод**: Таблица reviews состоит из  4 столбцов и 2793  строк.

## Задачи

### Задача 1. Посчитать, сколько книг вышло после 1 января 2000 года;


In [9]:
query = '''
SELECT COUNT (DISTINCT book_id)
FROM books
WHERE
    publication_date > '2000-01-01'
    
    '''

pd.io.sql.read_sql(query, con = engine)



count
0    819

**Вывод**: Всего вышло после 1 января 2000 года - 819 книг.

### Задача 2. Для каждой книги посчитайте количество обзоров и среднюю оценку;

**Вывод**: Самое большое кол-во обзоров имеет книга 948 для нее было сделано 1120 ревью, при этом средняя оценка не большая, всего 3.66.

In [11]:
query = '''

WITH re1 AS 

( SELECT 
book_id,
COUNT(review_id) AS count_reviews
FROM reviews
GROUP BY book_id)

SELECT
DISTINCT  b.book_id AS book,
b.title AS books_name,
re1.count_reviews AS count_reviews,
AVG (ra.rating ) AS avg_raiting

FROM books AS b
LEFT JOIN 
 re1
        
ON re1.book_id = b.book_id
        
LEFT JOIN ratings AS ra

ON ra.book_id = b.book_id
GROUP BY
    book,books_name, count_reviews
    
ORDER BY
   count_reviews DESC
   
   LIMIT 20;
   
   
  '''

pd.io.sql.read_sql(query, con = engine)


book                                         books_name  count_reviews  \
0     83    Anne Rice's The Vampire Lestat: A Graphic Novel            NaN   
1    387                               Leonardo's Notebooks            NaN   
2    221                          Essential Tales and Poems            NaN   
3    808                            The Natural Way to Draw            NaN   
4    672   The Cat in the Hat and Other Dr. Seuss Favorites            NaN   
5    191  Disney's Beauty and the Beast (A Little Golden...            NaN   
6    948                            Twilight (Twilight  #1)            7.0   
7    854                                           The Road            6.0   
8    673                             The Catcher in the Rye            6.0   
9    779  The Lightning Thief (Percy Jackson and the Oly...            6.0   
10   497                          Outlander (Outlander  #1)            6.0   
11   627                                      The Alchemist            6.0   
12   696             The Da Vinci Code (Robert Langdon  #2)            6.0   
13   207                                    Eat  Pray  Love            6.0   
14   750                The Hobbit  or There and Back Again            6.0   
15   963                                Water for Elephants            6.0   
16   299  Harry Potter and the Chamber of Secrets (Harry...            6.0   
17   302  Harry Potter and the Prisoner of Azkaban (Harr...            6.0   
18   656                                     The Book Thief            6.0   
19   695  The Curious Incident of the Dog in the Night-Time            6.0   

    avg_raiting  
0      3.666667  
1      4.000000  
2      4.000000  
3      3.000000  
4      5.000000  
5      4.000000  
6      3.662500  
7      3.772727  
8      3.825581  
9      4.080645  
10     4.125000  
11     3.789474  
12     3.830508  
13     3.395833  
14     4.125000  
15     3.977273  
16     4.287500  
17     4.414634  
18     4.264151  
19     4.081081

**Вывод**: у 6 книг нет ни одного обзора. У книги 948(Twilight (Twilight #1)) 7 обзоров и средний рейтинг 3.66. Максимальный средний рейтинг 5 баллов.

### Задача 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [12]:
query = '''

SELECT 
    COUNT(b.book_id) AS count_books,
    p.publisher AS publisher 

FROM books AS b
LEFT JOIN publishers AS p
ON b.publisher_id=p.publisher_id
WHERE num_pages>50

GROUP BY 
publisher

ORDER BY count_books DESC
LIMIT 3;


    '''

pd.io.sql.read_sql(query, con = engine)

count_books                 publisher
0           42             Penguin Books
1           31                   Vintage
2           25  Grand Central Publishing

**Вывод**: Издательство Penguin Books выпустило 42 книги толще 50 страниц.

### Задача 4.Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [15]:
query = '''

WITH ra AS
(        SELECT 
            book_id,
            AVG(rating) AS avg_rating,
            COUNT(rating_id) AS rating_cnt
        FROM ratings
        GROUP BY
        book_id
    )



SELECT
    a.author AS author,
    AVG(ra.avg_rating) AS avg_rating
FROM  books AS b
LEFT JOIN ra 
ON ra.book_id = b.book_id
LEFT JOIN authors AS a    
ON a.author_id = b.author_id

WHERE (rating_cnt) >= 50
GROUP BY a.author
ORDER BY
AVG(ra.avg_rating) DESC
LIMIT 5
'''
    

      

pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0         J.K. Rowling/Mary GrandPré    4.283844
1  Markus Zusak/Cao Xuân Việt Khương    4.264151
2                     J.R.R. Tolkien    4.258446
3                  Louisa May Alcott    4.192308
4                       Rick Riordan    4.080645

Автора с самой высокой средней оценкой книг  по заданому критирею явлется J.K. Rowling/Mary GrandPré	 имеет средний рейтинг 4.28

### Задача 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [17]:
query = '''


WITH
    user_rating AS (
    SELECT
        rat.username AS rat_username,
        COUNT(rating_id) AS ratings_count
    FROM
        ratings AS rat
    GROUP BY
        rat.username
    HAVING
        COUNT(rating_id) > 50
        ),
        
    user_review AS (
        SELECT
        rev.username AS rev_username,
     COUNT(review_id) AS reviews_count
        FROM
            reviews AS rev
        GROUP BY
            rev.username
    )
SELECT
  
AVG (reviews_count) AS AVG

FROM
user_rating
LEFT JOIN user_review 
ON user_rating.rat_username = user_review.rev_username
     
ORDER BY AVG (reviews_count)DESC


 Limit 5;
 ''' 
       
      

pd.io.sql.read_sql(query, con = engine)   

avg
0  24.333333

Среднее количество обзоров от пользователей потаваиших более 50 обзоров - 24,33.

## Общий вывод:

- Всего вышло после 1 января 2000 года - 819 книг.
-  у 6 книг нет ни одного обзора. У книги 948(Twilight (Twilight #1)) 7 обзоров и средний рейтинг 3.66. Максимальный средний рейтинг 5 баллов.
- Издательство Penguin Books выпустило 42 книги толще 50 страниц.
- Автора с самой высокой средней оценкой книг  по заданому критирею явлется J.K. Rowling/Mary GrandPré	 имеет средний рейтинг 4.28​.
- Среднее количество обзоров от пользователей потаваиших более 50 обзоров - 24,33.